### ISSUE TITLE
Question: Would there be support for Post Quantum Crypto? (Dilithium & Kyber)

### ISSUE THREAD

**Issue by Hussain1811** [opencryptoki/openssl-ibmca#117]
Will the IBM-CA provider (https://www.ibm.com/docs/en/linux-on-systems?topic=openssl-using-ibmca-provider) support post-quantum crypto algorithms in the future similar to the OQS provider (https://github.com/open-quantum-safe/oqs-provider) for Z and LinuxONE?

**Comment by ifranzki** [comment]
Currently it does not. 

As a first step the final versions of the post-quantum crypto algorithms need to be included into OpenSSL somehow. This could either be by adding support for those to the OpenSSL code base, or by including the OQS provider into the OpenSSL code base. It is up to the OpenSSL org to decide that.

Once this has happend, we can think about how to best HW accelerate post-quantum crypto algorithms. One option could be to enhance the IBMCA provider and add post quantum algorithms, another option could be to add HW acceleration right into the OpenSSL or OQS provider code. 

We haven't decided yet, but it is certainly on our wish list, too. 

I' leave this issue open for now as a feature request, and update once we there are any developments in that area.

**Comment by Hussain1811** [comment]
Now OpenSSL version 3.5.0 officially supports PQC algorithms (ML-DSA, ML-KEM, SLH-DSA)
Any path forward here now?

**Comment by ifranzki** [comment]
We still have this on our list of things to consider, but have not yet decided how exactly to pursue.

**Comment by Hussain1811** [comment]
Some probably related questions:
1)So OpenSSL can talk to both CCA and EP11 co-processors using the IBMCA engine/provider?
Confused as https://www.ibm.com/docs/en/linux-on-systems?topic=openssl-ibmca-in mentions that IBMCA is only for CCA
2)Is it possible to set this up on an x86 system where linux is running (EP11/CCA adapter being connected to this Linux on x86 host system)?

**Comment by ifranzki** [comment]
> 1)So OpenSSL can talk to both CCA and EP11 co-processors using the IBMCA engine/provider?

The IBMCA provider can only use crypto cards in Accelerator mode or in CCA mode. EP11 can not be used by IBMCA (EP11 does not support clear key crypto operations, as needed by IBMCA).

> 2)Is it possible to set this up on an x86 system where linux is running (EP11/CCA adapter being connected to this Linux on x86 host system)?

No, IBMCA is only available on Linux on IBM Z (s390x). From a hardware point of view one can use CCA crypto cards on x86 as well (CCA only!), but still IBMCA and libica (which it relies on) is only available on Linux on IBM Z.

**Comment by Hussain1811** [comment]
I see that the only way for OpenSSL to talk to EP11 is via the PKCS#11 engine (https://www.ibm.com/docs/en/linux-on-systems?topic=pkcs11-libp11-engine-use-case)? But this as well is only on Linux on Z

So, can IBMCA and libp11 be ported to x86?

**Comment by ifranzki** [comment]
> I see that the only way for OpenSSL to talk to EP11 is via the PKCS#11 engine

Right, actually today better use the PKCS#11 provider, see https://www.ibm.com/docs/en/linux-on-systems?topic=pkcs11-provider-use-case . Engines are deprecated by OpenSSL.

> But this as well is only on Linux on Z

The PKCS#11 provider is available for other platforms as well, just you need a PKCS#11 implementation on that platform. And yes, Opencryptoki's EP11 token is only available on Linux on IBM Z. 

> So, can IBMCA and libp11 be ported to x86?

IBMCA: no, its using libica which in turn is very platform specific. 
libp11: This should be available on other platforms already. But nowadays you better use the PKCS#11 provider (also available on other platforms already), not the engine, see above.

**Comment by Hussain1811** [comment]
Thank you for the references.

My goal is to connect OpenSSL to EP11 on a x86 RHEL host which has an IBM crypto card connected (configured in EP11 mode) via PCIe.
So, you mean to say this would be possible with https://github.com/opencryptoki/openssl-pkcs11-sign-provider  [OR] https://github.com/latchset/pkcs11-provider?

**Comment by ifranzki** [comment]
No, as far as I know EP11 can NOT be attached to x86 at all. Only crypto cards in CCA mode are supported for x86. 

But what you could do it to run Opencryptoki with its EP11 token on Linux on IBM Z system, and use the remote option of the p11-kit (https://www.ibm.com/docs/en/linux-on-systems?topic=scenarios-configuring-remote-pkcs-11-service) to make this Opencryptoki EP11 token available remotely, and use it from your x86 system. Then you can use OpenSSL with the PKCS#11 provider to connect through p11-kit to Opencryptoki and thus use the EP11 token.

**Comment by Hussain1811** [comment]
Thanks @ifranzki

**Comment by Hussain1811** [comment]
To configure PKCS11 provider for CCA, i have to specify the PKCS#11 driver in the openssl cnf file
`pkcs11-module-path = <CCA token.so>`

How can i build this CCA token?

**Comment by ifranzki** [comment]
No, you specify Opencryptoki's libopencryptoki.so in the openssl cnf file for `pkcs11-module-path`. Then in the PKCS#11-URI you use to reference the key you specify the token name of the CCA token.
You can find examples here: https://www.ibm.com/docs/en/linux-on-systems?topic=case-apache-http-server-configuration-pkcs11-provider

**Comment by Hussain1811** [comment]
Thank You
Does opencryptoki version 3.24.0 support CCA crypto cards on x86?

**Comment by ifranzki** [comment]
Yes: https://github.com/opencryptoki/opencryptoki/blob/91ecd8214aafd585a26cd6a6b8ab38885911c004/ChangeLog#L3
